File: Week4/Training/MVVMArchitectureComp/app/src/main/java/com/example/admin/mvvmarchitecturecomp/data/remote/RepositoryModule/RepositoryImpl.java
package com.example.admin.mvvmarchitecturecomp.data.remote.RepositoryModule;

import android.annotation.SuppressLint;
import android.arch.lifecycle.MutableLiveData;

import com.example.admin.mvvmarchitecturecomp.data.remote.RemoteServiceHelper;
import com.example.admin.mvvmarchitecturecomp.data.remote.models.ApiList;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class RepositoryImpl implements Repository {

    private MutableLiveData<List<ApiList>> weatherForecastLiveData = new MutableLiveData<>();
    private RemoteServiceHelper remoteServiceHelper;

    public RepositoryImpl(RemoteServiceHelper remoteServiceHelper){
        this.remoteServiceHelper = remoteServiceHelper;
    }

    @SuppressLint("CheckResult")
    @Override
    public void getWeatherForecast() {
        remoteServiceHelper.getWeatherForecast()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response ->
                    weatherForecastLiveData.setValue(response.getApiList())
                , Throwable::printStackTrace);
    }

    public MutableLiveData<List<ApiList>> getWeatherForecastLiveData() {
        return weatherForecastLiveData;
    }
}

File: Week5/Homework/CodeChallengeKotlin/app/src/main/java/com/example/admin/codechallengekotlin/data/repository/RepositoryImpl.kt
package com.example.admin.codechallengekotlin.data.repository

import android.arch.lifecycle.MutableLiveData
import android.util.Log
import com.example.admin.codechallengekotlin.data.remote.RemoteServiceHelper
import com.example.admin.codechallengekotlin.data.remote.models.MetaWeatherResponse
import kotlinx.coroutines.experimental.CoroutineExceptionHandler
import kotlinx.coroutines.experimental.launch
import retrofit2.HttpException

class RepositoryImpl(private val remoteServiceHelper: RemoteServiceHelper) : Repository {

    override val weatherForecastLiveData: MutableLiveData<List<MetaWeatherResponse>> = MutableLiveData()

    override fun getWeatherForecast(woeId: String) {

        val exceptionHandler = CoroutineExceptionHandler { _, throwable ->
            if (throwable is HttpException) {
                //TODO implement exception handling
                Log.d("Main", "In HttpException")

            } else {
                //TODO implement exception handling
                Log.d("Main", "Other error", throwable.fillInStackTrace())
            }
        }

        launch(exceptionHandler) {
            val metaWeatherResponse = remoteServiceHelper.getWeatherForecast(woeId).await()
            weatherForecastLiveData.postValue(metaWeatherResponse)
        }
    }
}